Userland/Libraries/LibGfx/VectorN.h

```diff
@@ -0,0 +1,263 @@
+#    pragma GCC unroll 4
+#endif
+        for (auto i = 0u; i < N; ++i) {
+          result.m_data[i] = -m_data[i];
+        }
+        return result;
+    }
+
+    constexpr VectorN operator/(const VectorN& other) const
+    {
+        VectorN result {};
+#ifdef __clang__
+#    pragma unroll 4
+#else
+#    pragma GCC unroll 4
+#endif
+        for (auto i = 0u; i < N; ++i) {
+            result.m_data[i] = m_data[i] / other.m_data[i];
+        }
+        return result;
+    }
+
+    template<typename U>
+    constexpr VectorN operator*(U f) const
+    {
+        VectorN result {};
+#ifdef __clang__
+#    pragma unroll 4
+#else
+#    pragma GCC unroll 4
+#endif
+        for (auto i = 0u; i < N; ++i) {
+            result.m_data[i] = m_data[i] * f;
+        }
+        return result;
+    }
+
+    template<typename U>
+    constexpr VectorN operator/(U f) const
+    {
+        VectorN result {};
+#ifdef __clang__
+#    pragma unroll 4
+#else
+#    pragma GCC unroll 4
+#endif
+        for (auto i = 0u; i < N; ++i) {
+            result.m_data[i] = m_data[i] / f;
+        }
+        return result;
+    }
+
+    constexpr T dot(const VectorN& other) const
+    {
+        T result {};
+        result += x() * other.x();
+        result += y() * other.y();
+        if constexpr (N >= 3) {
+            result += z() * other.z();
+        }
+        if constexpr (N >= 4) {
+            result += w() * other.w();
+        }
+        return result;
+    }
+
+    constexpr VectorN cross(const VectorN& other) const requires(N == 3)
+    {
+        return VectorN(
+            y() * other.z() - z() * other.y(),
+            z() * other.x() - x() * other.z(),
+            x() * other.y() - y() * other.x());
+    }
+
+    constexpr VectorN normalized() const
+    {
+        VectorN copy { *this };
+        copy.normalize();
+        return copy;
+    }
+
+    constexpr VectorN clamped(T m, T x) const
+    {
+        VectorN copy { *this };
+        copy.clamp(m, x);
+        return copy;
+    }
+
+    constexpr void clamp(T min_value, T max_value)
+    {
+#ifdef __clang__
+#    pragma unroll 4
+#else
+#    pragma GCC unroll 4
+#endif
+        for (auto i = 0u; i < N; ++i) {
+            m_data[i] = max(min_value, m_data[i]);
+            m_data[i] = min(max_value, m_data[i]);
+        }
+    }
+
+    constexpr void normalize()
+    {
+        T const inv_length = 1 / length();
+        operator*(inv_length);
+    }
+
+    constexpr T length() const
+    {
+        if constexpr (N == 2) {
+            return AK::hypot(m_data[0] * m_data[0] + m_data[1] * m_data[1]);
+        } else if constexpr (N == 3) {
+            return AK::sqrt(m_data[0] * m_data[0] + m_data[1] * m_data[1] + m_data[2] * m_data[2]);
+        } else {
+            return AK::sqrt(m_data[0] * m_data[0] + m_data[1] * m_data[1] + m_data[2] * m_data[2] + m_data[3] * m_data[3]);
+        }
+    }
+
+    constexpr VectorN<3, T> xyz() const requires(N == 4)
+    {
+        if constexpr (N == 4) {
```

Comment:
Why this check if the function already `requires` it?

Reply:
I wasn't sure it would work, but it does... Thanks for catching it.